### PR TITLE
Use sygus to synthesize/verify rewrite rules

### PR DESCRIPTION
**Pull request by ajreynol**
Adds new options.

(1) --sygus-rr-synth:  With this option, the sygus solver will report candidate rewrite rules with output of the form:
(candidate-rewrite t1 t2)
where t1 and t2 are enumerated by the input grammar, are equivalent under sampling but do not rewrite to the same term.

(2) --sygus-rr-verify:  With this option, the sygus solver will report unsound instances of the rewriter with output of the form:
(unsound-rewrite t1 t2)
where t1 and t2 are enumerated by the input grammar, are not equivalent under sampling but rewrite to the same term.

### CODE REVIEW

**Comment by ajreynol** [comment]
I also considered the names sygus-rr-suggest and sygus-rr-assure, although these are less catchy/harder to remember.

File: src/theory/quantifiers/sygus_sampler.h [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  void initialize(TermDbSygus* tds, Node f, unsigned nsamples);
+  /** register term n with this sampler database
+   *
+   * forceKeep is whether we wish to force that n is chosen as a representative
+   * value in the trie.
+   */
+  Node registerTerm(Node n, bool forceKeep = false);
+  /** is contiguous
+   *
+   * This returns whether n's free variables are a prefix of the list of
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_sampler.h]: Say that the free variables are ordered before you test whether they are a prefix.

File: src/theory/quantifiers/sygus_sampler.h [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+   * d_type_vars[Int] = { x, y }, then 0, x, x+y, y+x are contiguous but y is
+   * not. This is useful for excluding terms from consideration that are
+   * alpha-equivalent to others.
+   */
+  bool isContiguous(Node n);
+  /** is ordered
+   *
+   * This returns whether n's free variables are in order with respect to
+   * variables in d_type_vars for each type. For instance, if
+   * d_type_vars[Int] = { x, y }, then 0, x, x+y are contiguous but y and y+x
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_sampler.h]: are contiguous -> are ordered

File: src/theory/quantifiers/sygus_sampler.cpp [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+ **
+ ** \brief Implementation of sygus_sampler
+ **/
+
+#include "theory/quantifiers/sygus_sampler.h"
+
+#include "util/bitvector.h"
+#include "util/random.h"
+
+using namespace std;
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_sampler.cpp]: There namespaces are not being used.

File: src/theory/quantifiers/sygus_sampler.cpp [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    if (index == ntotal)
+    {
+      // lazy child holds the leaf data
+      if (lt->d_lazy_child.isNull() || forceKeep)
+      {
+        lt->d_lazy_child = n;
+      }
+      return lt->d_lazy_child;
+    }
+    else
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_sampler.cpp]: no-else-after-return

File: src/theory/quantifiers/sygus_sampler.cpp [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+      std::vector<Node> ex;
+      if (lt->d_children.empty())
+      {
+        if (lt->d_lazy_child.isNull())
+        {
+          // no one has been here, we are done
+          lt->d_lazy_child = n;
+          return lt->d_lazy_child;
+        }
+        else
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_sampler.cpp]: ditto

File: src/theory/quantifiers/sygus_sampler.cpp [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+          TypeNode tn = cur.getType();
+          // if this variable is out of order
+          if (itv->second != fvs[tn].size())
+          {
+            return false;
+          }
+          fvs[tn].push_back(cur);
+        }
+      }
+      unsigned nchildren = cur.getNumChildren();
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_sampler.cpp]: initialize the var in the ``for``

File: src/theory/quantifiers/sygus_sampler.h [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+   * A sampler can be invalid if sample points cannot be generated for a type
+   * of an argument to function f.
+   */
+  bool d_is_valid;
+  /**
+   * Compute the variables from the domain of d_var_index that occur in n,
+   * store these in the vector fvs.
+   */
+  void computeFreeVariables(Node n, std::vector<Node>& fvs);
+  /** get random value for a type */
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_sampler.h]: Improve this documentation. Say that you sample differently for different types etc

**Pull request merged**